components/raftstore/src/store/peer.rs

```diff
@@ -1764,57 +2019,156 @@ where
             }
         };
 
-        Some(CollectedReady::new(invoke_ctx, ready))
+        match &ready_res {
+            HandleReadyResult::SendIOTask | HandleReadyResult::Snapshot { .. } => {
+                if let Some(write_worker) = ctx.write_worker.as_mut() {
+                    write_worker.batch.add_write_task(task);
+                } else {
+                    self.write_router.send_write_msg(
+                        ctx,
+                        self.unpersisted_readies.back().map(|r| r.number),
+                        WriteMsg::WriteTask(task),
+                    );
+                }
+            }
+            _ => (),
+        }
+
+        Some(CollectedReady::new(ready_res, ready, has_new_entries))
     }
 
     pub fn post_raft_ready_append<T: Transport>(
         &mut self,
         ctx: &mut PollContext<EK, ER, T>,
-        invoke_ctx: InvokeContext,
-        ready: &mut Ready,
-    ) -> Option<ApplySnapResult> {
-        if invoke_ctx.has_snapshot() {
-            // When apply snapshot, there is no log applied and not compacted yet.
-            self.raft_log_size_hint = 0;
-            // Pause `read_progress` to prevent serving stale read while applying snapshot
-            self.read_progress.pause();
-        }
-
-        let apply_snap_result = self.mut_store().post_ready(invoke_ctx);
-        let has_msg = !ready.persisted_messages().is_empty();
-
-        if apply_snap_result.is_some() {
-            self.pending_messages = ready.take_persisted_messages();
-
-            // The peer may change from learner to voter after snapshot applied.
-            let peer = self
-                .region()
-                .get_peers()
-                .iter()
-                .find(|p| p.get_id() == self.peer.get_id())
-                .unwrap()
-                .clone();
-            if peer != self.peer {
```

Comment:
If it's updated after persisting, then the peer can be slightly different from the one in raft node. And in the past, messages are sent with updated peer info, now the messages are sent with stale peer info.

Though I don't find problems with current code space.

Reply:
Make sense. I refine the code of `post_raft_ready_append` and build raft messages before really sending them when these messages are saved in `SnapshotContext` or `unpersisted_readies`.